text_generation/causal_lm/cpp/speculative_decoding.cpp

```diff
@@ -0,0 +1,300 @@
+    auto old_tensor_data = tensor.data<float>();
+    auto shape = tensor.get_shape();
+    size_t size_1 = shape[1];
+    size_t old_seq_len = shape[2];
+    size_t size_2 = shape[3];
+    
+    auto new_tensor = ov::Tensor{ov::element::f32, {BATCH_SIZE, size_1, new_seq_len, size_2}};
+    auto new_tensor_data = new_tensor.data<float>();
+    
+    // todo: assume batch_size is always 1, if it's not 1 one more cycle is needed
+    for (size_t i = 0; i < size_1; ++i) {
+        for (size_t j = 0; j < new_seq_len; ++j) {
+            auto dst_ptr = new_tensor_data + new_seq_len * size_2 * i +  size_2 * j;
+            auto src_ptr = old_tensor_data + old_seq_len * size_2 * i +  size_2 * j;
+            std::memcpy(dst_ptr, src_ptr, size_2 * sizeof(float));
+        }
+    }
+    return new_tensor;
+}
+
+void update_kv_cache(ov::InferRequest request, uint64_t new_seq_len) {
+    // trims kv_cache values up to the seq_len
+    auto num_outputs = request.get_compiled_model().outputs().size();
+    for (size_t idx = 1; idx < num_outputs; ++idx) {
+        auto old_tensor = request.get_output_tensor(idx);
+        auto trimmed_tensor = trimm_tensor(old_tensor, new_seq_len);
+        request.set_output_tensor(idx, trimmed_tensor);
+    }
+}
+
+
+auto tokens_to_string(ov::InferRequest& detokenizer, std::vector<int64_t> tokens) {
+    return detokenize(detokenizer, tokens);
+}
+
+int main(int argc, char* argv[]) try {
+    if (argc != 4) {
+        throw std::runtime_error(std::string{"Usage: "} + argv[0] + " <DRAFT MODEL_DIR> <MAIN MODEL_DIR> '<PROMPT>'");
+    }
+
+    // Compile models
+    ov::Core core;
+    core.add_extension(OPENVINO_TOKENIZERS_PATH);  // OPENVINO_TOKENIZERS_PATH is defined in CMakeLists.txt
+    // tokenizer and detokenizer work on CPU only
+    ov::InferRequest tokenizer = core.compile_model(
+        std::string{argv[1]} + "/openvino_tokenizer.xml", "CPU").create_infer_request();
+    auto [input_ids, attention_mask] = tokenize(tokenizer, argv[3]);
+    ov::InferRequest detokenizer = core.compile_model(
+        std::string{argv[1]} + "/openvino_detokenizer.xml", "CPU").create_infer_request();
+    // todo: for debug reason add one more detokenizer since the original one is moved to TextStreamer
+    ov::InferRequest detokenizer_2 = core.compile_model(
+        std::string{argv[1]} + "/openvino_detokenizer.xml", "CPU").create_infer_request();
+    
+    // draft model
+    ov::InferRequest draft_model = core.compile_model(std::string{argv[1]} + "/openvino_model.xml", "CPU").create_infer_request();
+
+    draft_model.set_tensor("input_ids", input_ids);
+    draft_model.set_tensor("attention_mask", attention_mask);
+    ov::Tensor position_ids = draft_model.get_tensor("position_ids");
+    position_ids.set_shape(input_ids.get_shape());
+    std::iota(position_ids.data<int64_t>(), position_ids.data<int64_t>() + position_ids.get_size(), 0);
+    init_key_values(draft_model);
+    draft_model.infer();
+
```

Comment:
I also propose running the cpp linter for the code and autotuning the style.

Reply:
done, removed a couple of warnings.